File: hub/agentlink.py
import os
import json
import sys
import requests
import threading
import time
import subprocess
import argparse
import textwrap
from rich.console import Console
from rich.panel import Panel
from rich.text import Text
from rich.align import Align
from rich.table import Table
from prompt_toolkit import PromptSession, print_formatted_text
from prompt_toolkit.patch_stdout import patch_stdout
from prompt_toolkit.formatted_text import HTML
from prompt_toolkit.key_binding import KeyBindings
from prompt_toolkit.history import FileHistory
from prompt_toolkit.styles import Style as PromptStyle

console = Console()
REGISTRY_PATH = "/root/AgentLink/hub/registry.json"
API_TOKEN = "agentlink_secret_123"

# Institutional Palette
C_PRIMARY = "#5f87af" # Steel Blue
C_SECONDARY = "#ffaf00" # Industrial Amber
C_ACCENT = "#87af87" # Sage Green
C_DIM = "#444444"
C_WHITE = "#eeeeee"

def get_headers():
    return {"Authorization": f"Bearer {API_TOKEN}"}

def load_registry():
    if not os.path.exists(REGISTRY_PATH): return []
    with open(REGISTRY_PATH, 'r') as f:
        try: return json.load(f)
        except: return []

def get_ascii_header():
    width = console.width
    
    # Static Pixel-Perfect Components
    # Bot: 19 chars
    b_lines = [
        r"     .───────.     ",
        r"    o| o   o |o    ",
        r"     |  ───  |     ",
        r"     '───────'     ",
        r"      /─────\      "
    ]
    
    # Agent: 44 chars
    a_lines = [
        " █████   ██████  ███████ ███    ██ ████████ ",
        "██   ██ ██       ██      ████   ██    ██    ",
        "███████ ██   ███ █████   ██ ██  ██    ██    ",
        "██   ██ ██    ██ ██      ██  ██ ██    ██    ",
        "██   ██  ██████  ███████ ██   ████    ██    "
    ]
    
    # Link: 33 chars
    l_lines = [
        "██      ██ ███    ██ ██   ██     ",
        "██      ██ ████   ██ ██  ██      ",
        "██      ██ ██ ██  ██ █████       ",
        "██      ██ ██  ██ ██ ██  ██      ",
        "███████ ██ ██   ████ ██   ██     "
    ]

    # Tier 1: Ultra-Compact (< 80 chars)
    if width < 80:
        header_text = Text.from_markup(f"[bold {C_SECONDARY}]〈[ o ─-─ o ]〉[/] [bold {C_PRIMARY}]AGENT[/][bold {C_ACCENT}]LINK[/]")
        return Align.center(Panel(header_text, border_style=C_PRIMARY, padding=(0, 2)))

    # Tier 2: Mid-Size (Logo Only, 80-120 chars)
    if width < 125:
        mid_content = []
        for i in range(5):
            line = Text()
            line.append(a_lines[i], style=f"bold {C_PRIMARY}")
            line.append(l_lines[i], style=f"bold {C_ACCENT}")
            mid_content.append(line)
        return Align.center(Text.assemble(*[ln + "\n" for ln in mid_content]))

    # Tier 3: Full Industrial (Bots + Logo, > 125 chars)
    # Using a Table Grid to lock alignment and prevent slanting
    grid = Table.grid(expand=False)
    grid.add_column(width=21, justify="right") # Left Bot
    grid.add_column(width=79, justify="center") # Logo
    grid.add_column(width=21, justify="left") # Right Bot

    for i in range(5):
        bot_l = Text(b_lines[i], style=f"bold {C_SECONDARY}")
        
        logo = Text()
        logo.append(a_lines[i], style=f"bold {C_PRIMARY}")
        logo.append(l_lines[i], style=f"bold {C_ACCENT}")
        
        bot_r = Text(b_lines[i], style=f"bold {C_SECONDARY}")
        
        grid.add_row(bot_l, logo, bot_r)

    return Align.center(grid)

def list_servers():
    registry = load_registry()
    table = Table(box=None, header_style=f"bold {C_SECONDARY}")
    table.add_column("NODE IDENTITY", style=C_PRIMARY, width=22)
    table.add_column("IPV4 ADDRESS", style="white", width=15)
    table.add_column("LINK STATUS", width=15)
    for node in registry:
        try:
            r = requests.get(f"http://{node['ip']}:8092/status", headers=get_headers(), timeout=1)
            status = f"[bold {C_ACCENT}]● OPERATIONAL[/bold {C_ACCENT}]" if r.status_code == 200 else "[bold red]○ DISCONNECTED[/bold red]"
        except: status = "[bold red]○ DISCONNECTED[/bold red]"
        table.add_row(node['name'].upper(), node['ip'], status)
    return Panel(table, title=f"[bold {C_WHITE}]Managed Fleet Core[/bold {C_WHITE}]", border_style=C_PRIMARY, expand=False)

def posses_server(name):
    registry = load_registry()
    node = next((n for n in registry if n['name'] == name), None)
    if not node: return
    ip = node['ip']
    console.clear()
    console.print(get_ascii_header())
    
    box_width = min(console.width - 2, 124)
    inner_w = box_width - 6
    
    # Session Header
    console.print(Align.center(Panel(
        Align.center(f"[bold {C_WHITE}]SECURE UPLINK:[/bold {C_WHITE}] [bold {C_PRIMARY}]{name.upper()}[/bold {C_PRIMARY}] @ {ip}\n[dim]Protocol: Neural-PTY | Encryption: Industrial Strength[/dim]"),
        border_style=C_ACCENT,
        title=f"[bold {C_ACCENT}]SESSION INITIALIZED[/bold {C_ACCENT}]",
        width=box_width
    )))

    stop_event = threading.Event()
    state = {"active_model": "gemini", "is_started": False}

    def boxed_print(text, color=C_WHITE):
        for line in text.splitlines():
            segments = textwrap.wrap(line, width=inner_w) if line.strip() else [""]
            for segment in segments:
                padding = " " * (inner_w - len(segment))
                print_formatted_text(HTML(f'<style fg="{C_ACCENT}"> │ </style> <style fg="{color}">{segment}</style>{padding} <style fg="{C_ACCENT}">│</style>'))

    def log_tailer():
        last_logs = ""
        while not stop_event.is_set():
            try:
                status_r = requests.get(f"http://{ip}:8092/status", headers=get_headers(), timeout=1)
                if status_r.status_code == 200:
                    data = status_r.json()
                    state["active_model"] = data.get("active_model", "gemini")
                    state["is_started"] = data.get("started", False)
                if state["is_started"]:
                    r = requests.get(f"http://{ip}:8092/logs", headers=get_headers(), timeout=2)
                    if r.status_code == 200:
                        new_logs = r.text
                        if new_logs != last_logs:
                            diff = new_logs[len(last_logs):]
                            if diff: boxed_print(diff.strip("\n"))
                            last_logs = new_logs
            except: pass
            time.sleep(0.5)

    threading.Thread(target=log_tailer, daemon=True).start()

    kb = KeyBindings()
    @kb.add('escape')
    def _(event):
        boxed_print("CMDS: /start, /switch <model>, /clear, /exit", C_SECONDARY)

    history_file = f"/root/.agentlink_history_{name}"
    custom_style = PromptStyle.from_dict({
        'prompt': f'{C_PRIMARY} bold',
        'toolbar': f'bg:{C_DIM} {C_WHITE} bold',
        'status-on': f'{C_ACCENT} bold',
        'status-off': 'red bold',
    })

    session = PromptSession(key_bindings=kb, history=FileHistory(history_file), style=custom_style)

    def bottom_toolbar():
        status_tag = "status-on" if state["is_started"] else "status-off"
        status_text = "ACTIVE" if state["is_started"] else "STANDBY"
        return HTML(f' <b>NODE:</b> {name.upper()} | <b>CORE:</b> {state["active_model"].upper()} | <b>LINK:</b> <{status_tag}>{status_text}</{status_tag}> | [Esc] Help')

    try:
        with patch_stdout():
            while True:
                # Continuous Box
                h_line = "─" * (box_width - 2)
                print_formatted_text(HTML(f'<style fg="{C_ACCENT}">╭{h_line}╮</style>'))
                
                ui_input = session.prompt(
                    HTML(f'<style fg="{C_ACCENT}"> │ </style><style fg="{C_PRIMARY}"><b> > </b></style>'), 
                    bottom_toolbar=bottom_toolbar
                )
                
                print_formatted_text(HTML(f'<style fg="{C_ACCENT}">╰{h_line}╯</style>'))
                
                if not ui_input: continue
                if ui_input == "/exit": break
                elif ui_input == "/start":
                    r = requests.post(f"http://{ip}:8092/start", headers=get_headers())
                    boxed_print("SYSTEM: Neural uplink operational.", C_ACCENT)
                elif ui_input == "/clear":
                    console.clear()
                    console.print(get_ascii_header())
                elif ui_input == "/help":
                    boxed_print("CMDS: /start, /switch <model>, /clear, /exit", C_SECONDARY)
                elif ui_input.startswith("/switch "):
                    m = ui_input.split(" ")[1]
                    requests.post(f"http://{ip}:8092/switch", json={"model": m}, headers=get_headers())
                    boxed_print(f"SYSTEM: Routing to {m.upper()}...", C_SECONDARY)
                else:
                    if state["is_started"]:
                        requests.post(f"http://{ip}:8092/prompt", json={"text": ui_input}, headers=get_headers())
                    else:
                        boxed_print("ERROR: Link standby. Execute /start.", "red")
    except (KeyboardInterrupt, EOFError):
        pass
    finally:
        stop_event.set()

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("target", nargs="?")
    args = parser.parse_args()
    registry = load_registry()
    node_names = [n['name'] for n in registry]

    if not args.target:
        console.clear()
        console.print(get_ascii_header())
        console.print("")
        tips = f"[bold {C_PRIMARY}]1.[/] Access Node: agentlink <name>\n[bold {C_PRIMARY}]2.[/] Wake Logic: /start\n[bold {C_PRIMARY}]3.[/] Route Core: /switch qwen"
        console.print(Align.center(Panel(tips, title=f"[bold {C_SECONDARY}]Quick Procedures[/]", border_style=C_SECONDARY, expand=False)))
        console.print(Align.center(list_servers()))
    elif args.target == "list":
        console.print(get_ascii_header())
        console.print(Align.center(list_servers()))
    elif args.target in node_names:
        posses_server(args.target)
    else:
        console.print(f"[bold red]FATAL:[/bold red] Identity '{args.target}' not recognized.")
        console.print(Align.center(list_servers()))

File: hub/agentlink.js
#!/usr/bin/env node
"use strict";

import React, { useState, useEffect, useRef } from 'react';
import { render, Box, Text, useApp, useStdout, useInput } from 'ink';
import TextInput from 'ink-text-input';
import SelectInput from 'ink-select-input';
import axios from 'axios';
import fs from 'fs';
import chalk from 'chalk';
import { Command } from 'commander';
import { loadRegistry, enrollNode, Vault, loadVault, Sanitizer, deleteNode, updateNode } from './fleet_manager.js';

const h = React.createElement;
const API_TOKEN = "agentlink_secret_123";
const C_PRIMARY = "#5f87af"; 
const C_SECONDARY = "#ffaf00"; 
const C_ACCENT = "#87af87"; 
const C_WHITE = "#eeeeee";
const C_DIM = "#444444";

function getHeaders() { return { "Authorization": `Bearer ${API_TOKEN}` }; }

const Header = () => {
    const b_lines = [".───────.", "o| o   o |o", " |  ───  | ", "'───────'", String.raw` /─────\ `];
    const a_lines = [" █████   ██████  ███████ ███    ██ ████████ ", "██   ██ ██       ██      ████   ██    ██    ", "███████ ██   ███ █████   ██ ██  ██    ██    ", "██   ██ ██    ██ ██      ██  ██ ██    ██    ", "██   ██  ██████  ███████ ██   ████    ██    "];
    const l_lines = ["██      ██ ███    ██ ██   ██     ", "██      ██ ████   ██ ██  ██      ", "██      ██ ██ ██  ██ █████       ", "██      ██ ██  ██ ██ ██  ██      ", "███████ ██ ██   ████ ██   ██     "];
    return h(Box, { flexDirection: "row", justifyContent: "center", width: "100%", marginBottom: 1 },
        h(Box, { flexDirection: "column", alignItems: "center" }, b_lines.map((l, i) => h(Text, { key: `lb-${i}`, color: C_SECONDARY, bold: true }, l))),
        h(Box, { flexDirection: "column", marginX: 3 }, a_lines.map((l, i) => h(Box, { key: `logo-${i}` }, h(Text, { color: C_PRIMARY, bold: true }, l), h(Text, { color: C_ACCENT, bold: true }, l_lines[i])))),
        h(Box, { flexDirection: "column", alignItems: "center" }, b_lines.map((l, i) => h(Text, { key: `rb-${i}`, color: C_SECONDARY, bold: true }, l)))
    );
};

const BlockRenderer = ({ block }) => {
    const content = Sanitizer.scrub(block.content);
    switch (block.type) {
        case 'USER': return h(Box, { marginTop: 1 }, h(Text, { color: C_SECONDARY, bold: true }, `✦ User: `), h(Text, { color: C_WHITE }, content));
        case 'SYSTEM': return h(Box, { marginY: 0 }, h(Text, { color: C_DIM, italic: true }, `  ${content}`));
        case 'ACTION': return h(Box, { borderStyle: "round", borderColor: C_PRIMARY, paddingX: 1, marginY: 1, width: "100%" }, h(Text, { color: C_ACCENT }, content));
        default: return h(Box, { paddingLeft: 2, width: "100%" }, h(Text, {}, block.isPending ? content + "█" : content));
    }
};

const CommandDropdown = ({ items, onSelect }) => {
    const [selectedIndex, setSelectedIndex] = useState(0);
    useInput((input, key) => {
        if (key.upArrow) setSelectedIndex(Math.max(0, selectedIndex - 1));
        if (key.downArrow) setSelectedIndex(Math.min(items.length - 1, selectedIndex + 1));
        if (key.return) onSelect(items[selectedIndex]);
    });
    return h(Box, { flexDirection: "column", borderStyle: "round", borderColor: C_DIM, paddingX: 1, width: 100, backgroundColor: "#111111", marginBottom: 0 },
        items.map((item, i) => (
            h(Box, { key: item.value, width: "100%" },
                h(Text, { color: i === selectedIndex ? C_WHITE : C_DIM, bold: i === selectedIndex }, i === selectedIndex ? "● " : "  "),
                h(Box, { width: 12 }, h(Text, { color: i === selectedIndex ? C_PRIMARY : C_DIM }, item.value.slice(1))),
                h(Text, { color: i === selectedIndex ? C_WHITE : C_DIM }, item.description)
            )
        )),
        h(Box, { justifyContent: "space-between", marginTop: 0 },
            h(Text, { color: C_DIM }, " ▼ "),
            h(Text, { color: C_DIM }, `(${selectedIndex + 1}/${items.length})`)
        )
    );
};

const Possession = ({ node, onExitToDashboard, onTriggerWorkflow }) => {
    const { exit } = useApp();
    const [input, setInput] = useState('');
    const [blocks, setBlocks] = useState([]);
    const [state, setState] = useState({ authenticated: false, pending_ssh: null, yolo_mode: false, active_model: 'gemini', sub_model: 'pro', available_models: [] });
    const [pendingAuth, setPendingAuth] = useState(false);
    const [showMenu, setShowMenu] = useState(null); 

    const { stdout } = useStdout();
    const termHeight = stdout ? stdout.rows : 40;

    const allCommands = [
        { label: 'yolo', value: '/yolo', description: '💀 Toggle Autonomous Mode (Skip all security gates)' },
        { label: 'auth', value: '/auth', description: 'Authenticate with AI accounts' },
        { label: 'setup', value: '/setup', description: 'Configure system credentials' },
        { label: 'switch', value: '/switch', description: 'Switch between AI cores' },
        { label: 'model', value: '/model', description: 'Select model version' },
        { label: 'enroll', value: '/enroll', description: 'Register a remote server' },
        { label: 'edit', value: '/edit', description: 'Modify server identity' },
        { label: 'clear', value: '/clear', description: 'Clear history and purge buffers' },
        { label: 'help', value: '/help', description: 'Show all commands' },
        { label: 'exit', value: '/exit', description: 'Return to dashboard' }
    ];

    useEffect(() => {
        axios.post(`http://${node.ip}:8092/reset`, {}, { headers: getHeaders() });
        const interval = setInterval(async () => {
            try {
                const statusR = await axios.get(`http://${node.ip}:8092/status`, { headers: getHeaders(), timeout: 1000 });
                const s = statusR.data;
                setState(s);

                // AUTO-APPROVAL LOGIC: If YOLO is on, automatically send vault credentials
                if (s.yolo_mode && s.pending_ssh) {
                    const vault = loadVault();
                    const creds = vault[s.pending_ssh];
                    const registry = loadRegistry();
                    const targetNode = registry.find(n => n.name === s.pending_ssh);
                    if (creds && targetNode) {
                        await axios.post(`http://${node.ip}:8092/ssh/authorize`, {
                            node_name: s.pending_ssh, user: creds.user, pass: Vault.decrypt(creds.pass), ip: targetNode.ip
                        }, { headers: getHeaders() });
                    }
                } else if (s.pending_ssh) {
                    setPendingAuth(true);
                }

                const logsR = await axios.get(`http://${node.ip}:8092/logs`, { headers: getHeaders(), timeout: 1000 });
                setBlocks(logsR.data.slice(-(termHeight - 18)));
            } catch (e) {}
        }, 400); 
        return () => clearInterval(interval);
    }, [node.ip, termHeight]);

    const handleSelect = async (item) => {
        setShowMenu(null);
        try {
            if (item.type === 'core') await axios.post(`http://${node.ip}:8092/switch`, { model: item.value }, { headers: getHeaders() });
            else if (item.type === 'model') await axios.post(`http://${node.ip}:8092/switch`, { model: state.active_model, subModel: item.value }, { headers: getHeaders() });
        } catch (e) {}
    };

    useInput(async (input_key, key) => {
        if (pendingAuth && key.return) {
            const vault = loadVault();
            const nodeName = state.pending_ssh;
            const creds = vault[nodeName];
            const registry = loadRegistry();
            const targetNode = registry.find(n => n.name === nodeName);
            if (creds && targetNode) {
                await axios.post(`http://${node.ip}:8092/ssh/authorize`, { node_name: nodeName, user: creds.user, pass: Vault.decrypt(creds.pass), ip: targetNode.ip }, { headers: getHeaders() });
                setPendingAuth(false);
            }
        }
        if (key.escape) { setPendingAuth(false); setShowMenu(null); setSetupStep(null); }
    });

    const handleSubmit = async (value) => {
        const cmd = value.trim();
        if (!cmd) return;
        setInput('');

        if (cmd === '/yolo') {
            await axios.post(`http://${node.ip}:8092/yolo`, { enabled: !state.yolo_mode }, { headers: getHeaders() });
            return;
        }
        if (cmd === '/exit') onExitToDashboard();
        if (cmd === '/switch') { setShowMenu('cores'); return; }
        if (cmd === '/model') { setShowMenu('models'); return; }
        if (cmd === '/setup') { setSetupStep('google_key'); return; }
        if (cmd === '/enroll') { onTriggerWorkflow('enroll'); return; }
        if (cmd === '/edit') { onTriggerWorkflow('edit'); return; }
        if (cmd === '/delete') { onTriggerWorkflow('delete'); return; }
        if (cmd === '/clear') { await axios.post(`http://${node.ip}:8092/reset`, {}, { headers: getHeaders() }); return; }
        
        if (cmd === '/help') {
            await axios.post(`http://${node.ip}:8092/log`, { message: "📖 COMMAND DEFINITIONS:", origin: "SYSTEM" }, { headers: getHeaders() });
            for (const c of allCommands) await axios.post(`http://${node.ip}:8092/log`, { message: `  ${chalk.bold(c.value.padEnd(10))} ${c.description}`, origin: "SYSTEM" }, { headers: getHeaders() });
            return;
        }
        try { await axios.post(`http://${node.ip}:8092/prompt`, { text: cmd }, { headers: getHeaders() }); } catch (e) {}
    };

    const suggestions = input.startsWith('/') && !showMenu ? allCommands.filter(c => c.value.startsWith(input)) : [];

    return h(Box, { flexDirection: "column", paddingX: 4, flexGrow: 1, minHeight: termHeight - 2 },
        h(Header),
        h(Box, { flexDirection: "column", flexGrow: 1, marginBottom: 1, width: "100%" },
            blocks.map((b) => h(BlockRenderer, { key: b.id, block: b }))
        ),
        suggestions.length > 0 && h(CommandDropdown, { items: suggestions, onSelect: (item) => { setInput(item.value + ' '); } }),
        pendingAuth && h(Box, { borderStyle: "double", borderColor: "red", padding: 2, flexDirection: "column", alignSelf: "center", backgroundColor: "#111111", width: 80, marginBottom: 1 },
            h(Text, { bold: true, color: "red" }, "⚠  SECURITY: REMOTE ACCESS REQUEST"),
            h(Box, { marginTop: 1 }, h(Text, { color: C_WHITE }, `The agent is requesting to inhabit `), h(Text, { color: C_PRIMARY, bold: true }, state.pending_ssh?.toUpperCase())),
            h(Text, { color: C_WHITE, marginTop: 1 }, "Authorize this session using encrypted vault credentials?"),
            h(Box, { marginTop: 1 }, h(Text, { color: C_ACCENT, bold: true }, "Press [ENTER] to Authorize"), h(Text, { color: C_DIM }, " | "), h(Text, { color: "red", bold: true }, "[ESC] to Deny"))
        ),
        h(Box, { borderStyle: "classic", borderColor: C_PRIMARY, paddingX: 1, marginBottom: 1, flexDirection: "row" },
            h(Text, { color: C_PRIMARY, bold: true }, " │ > "),
            h(Box, { flexGrow: 1 }, h(TextInput, { value: input, onChange: setInput, onSubmit: handleSubmit }))
        ),
        h(Box, { paddingX: 1, justifyContent: "space-between" },
            h(Box, {}, h(Text, { dimColor: true }, " ~   "), h(Text, { color: C_ACCENT }, "no sandbox")),
            h(Box, {}, 
                h(Text, { color: state.yolo_mode ? C_SECONDARY : C_WHITE, bold: state.yolo_mode }, state.yolo_mode ? "💀 YOLO ACTIVE" : " PROTECTED"),
                h(Text, { dimColor: true }, " | "),
                h(Text, { color: C_SECONDARY, bold: true }, state.active_model?.toUpperCase()),
                h(Text, { dimColor: true }, " | "), 
                h(Text, { color: state.authenticated ? C_ACCENT : 'red', bold: true }, state.authenticated ? " AUTH" : "🔑 NO KEY")
            )
        )
    );
};

const App = ({ target }) => {
    const { exit } = useApp();
    const [nodes, setNodes] = useState([]);
    const [possessionNode, setPossessionNode] = useState(null);
    const [enrollStep, setEnrollStep] = useState(null); 
    const [enrollValue, setEnrollValue] = useState('');
    const [temp, setTemp] = useState({});
    const [dashboardInput, setDashboardInput] = useState('');
    const [workflow, setWorkflow] = useState({ type: null, node: null, field: null });

    const refreshNodes = async () => {
        const reg = loadRegistry();
        const updated = await Promise.all(reg.map(async (node) => {
            try { const r = await axios.get(`http://${node.ip}:8092/status`, { headers: getHeaders(), timeout: 1000 }); return { ...node, status: 'OPERATIONAL' }; }
            catch (e) { return { ...node, status: 'DISCONNECTED' }; }
        }));
        setNodes(updated);
    };

    useEffect(() => {
        const reg = loadRegistry();
        if (target && target !== 'list') {
            const node = reg.find(n => n.name.toLowerCase() === target.toLowerCase());
            if (node) setPossessionNode(node);
        }
        refreshNodes();
    }, [target]);

    useEffect(() => {
        if (!possessionNode && !enrollStep && !workflow.type) {
            const timer = setInterval(refreshNodes, 5000);
            return () => clearInterval(timer);
        }
    }, [possessionNode, enrollStep, workflow]);

    const handleEnroll = async (val) => {
        const steps = ['name', 'ip', 'user', 'pass', 'done'];
        const currentIdx = steps.indexOf(enrollStep);
        const next = steps[currentIdx + 1];
        const newTemp = { ...temp, [enrollStep]: val };
        setTemp(newTemp);
        setEnrollValue('');
        if (next === 'done') {
            await enrollNode(newTemp.ip, newTemp.name, newTemp.user, newTemp.pass);
            setEnrollStep(null);
            refreshNodes();
        } else setEnrollStep(next);
    };

    const handleDashboardSubmit = (cmd) => {
        setDashboardInput('');
        if (cmd === 'enroll' || cmd === '/enroll') setEnrollStep('name');
        else if (cmd === 'delete' || cmd === '/delete') setWorkflow({ type: 'delete', node: null, field: null });
        else if (cmd === 'edit' || cmd === '/edit') setWorkflow({ type: 'edit', node: null, field: null });
        else if (cmd === 'exit' || cmd === '/exit') exit();
        else {
            const node = nodes.find(n => n.name.toLowerCase() === cmd.toLowerCase());
            if (node) setPossessionNode(node);
        }
    };

    if (possessionNode) return h(Possession, { 
        node: possessionNode, 
        onExitToDashboard: () => setPossessionNode(null),
        onTriggerWorkflow: (type) => { setPossessionNode(null); if (type === 'enroll') setEnrollStep('name'); else setWorkflow({ type, node: null, field: null }); }
    });

    return h(Box, { flexDirection: "column", padding: 2 },
        h(Header),
        enrollStep ? (
            h(Box, { flexDirection: "column", alignSelf: "center", borderStyle: "double", borderColor: C_SECONDARY, padding: 2, width: 65 },
                h(Text, { bold: true, color: C_SECONDARY }, "🚀 SECURE FLEET ENROLLMENT"),
                h(Text, { color: C_WHITE, marginTop: 1 }, enrollStep === 'name' ? "1. Give this server a name:" : enrollStep === 'ip' ? "2. IPv4 Address:" : enrollStep === 'user' ? "3. Username:" : "4. Password:"),
                h(Box, { borderStyle: "single", borderColor: C_DIM, marginTop: 1, paddingX: 1 }, h(TextInput, { value: enrollValue, onChange: setEnrollValue, onSubmit: handleEnroll, mask: enrollStep === 'pass' ? '*' : '' })),
                h(Text, { color: C_DIM, marginTop: 1 }, "Press [Enter] to continue")
            )
        ) : workflow.type === 'delete' ? (
            h(Box, { flexDirection: "column", alignSelf: "center", borderStyle: "double", borderColor: "red", padding: 2, width: 65 },
                h(Text, { bold: true, color: "red" }, "🗑 DELETE SERVER"),
                h(SelectInput, { items: nodes.filter(n => n.name !== 'local-node').map(n => ({ label: n.name.toUpperCase(), value: n.name })), onSelect: (item) => { deleteNode(item.value); setWorkflow({ type: null }); refreshNodes(); } }),
                h(Text, { color: C_DIM, marginTop: 1 }, "[Esc] to cancel")
            )
        ) : workflow.type === 'edit' && !workflow.node ? (
            h(Box, { flexDirection: "column", alignSelf: "center", borderStyle: "double", borderColor: C_PRIMARY, padding: 2, width: 65 },
                h(Text, { bold: true, color: C_PRIMARY }, "🔧 SELECT TO EDIT"),
                h(SelectInput, { items: nodes.filter(n => n.name !== 'local-node').map(n => ({ label: n.name.toUpperCase(), value: n.name })), onSelect: (item) => setWorkflow({ ...workflow, node: item.value }) }),
                h(Text, { color: C_DIM, marginTop: 1 }, "[Esc] to cancel")
            )
        ) : workflow.type === 'edit' && workflow.node && !workflow.field ? (
            h(Box, { flexDirection: "column", alignSelf: "center", borderStyle: "double", borderColor: C_PRIMARY, padding: 2, width: 65 },
                h(Text, { bold: true, color: C_PRIMARY }, `🔧 EDIT: ${workflow.node.toUpperCase()}`),
                h(SelectInput, { items: [{ label: '🏷 Rename', value: 'name' }, { label: '🌐 IP', value: 'ip' }, { label: ' User', value: 'user' }, { label: '🔑 Pass', value: 'pass' }], onSelect: (item) => setWorkflow({ ...workflow, field: item.value }) }),
                h(Text, { color: C_DIM, marginTop: 1 }, "Select field")
            )
        ) : workflow.type === 'edit' && workflow.field ? (
            h(Box, { flexDirection: "column", alignSelf: "center", borderStyle: "double", borderColor: C_ACCENT, padding: 2, width: 65 },
                h(Text, { bold: true, color: C_ACCENT }, ` UPDATE ${workflow.field.toUpperCase()}`),
                h(Box, { borderStyle: "single", borderColor: C_DIM, marginTop: 1, paddingX: 1 }, h(TextInput, { value: enrollValue, onChange: setEnrollValue, onSubmit: (val) => { updateNode(workflow.node, workflow.field, val); setWorkflow({ type: null }); setEnrollValue(''); refreshNodes(); } })),
                h(Text, { color: C_DIM, marginTop: 1 }, "Press [Enter] to save")
            )
        ) : (
            h(Box, { flexDirection: "column", alignSelf: "center" },
                h(Box, { flexDirection: "column", borderStyle: "round", borderColor: C_PRIMARY, padding: 1, width: 90 },
                    h(Box, { marginBottom: 1, justifyContent: "center" }, h(Text, { bold: true, color: C_SECONDARY }, "FLEET INTELLIGENCE COMMAND")),
                    nodes.map((node, i) => (
                        h(Box, { key: i },
                            h(Box, { width: 30 }, h(Text, { color: C_PRIMARY }, node.name.toUpperCase())),
                            h(Box, { width: 25 }, h(Text, { color: "white" }, node.ip)),
                            h(Box, {}, h(Text, { color: node.status === 'OPERATIONAL' ? C_ACCENT : 'red' }, node.status === 'OPERATIONAL' ? '● OPERATIONAL' : '○ DISCONNECTED'))
                        )
                    )),
                    h(Box, { borderStyle: "classic", borderColor: C_DIM, paddingX: 1, marginTop: 1 },
                        h(Text, { color: C_PRIMARY }, " COMMAND > "),
                        h(TextInput, { value: dashboardInput, onChange: setDashboardInput, onSubmit: handleDashboardSubmit, placeholder: "Type node name, 'edit', 'delete', or 'enroll'" })
                    )
                )
            )
        )
    );
};

const program = new Command();
program.arguments('[target]').action((target) => render(h(App, { target: target })));
program.parse(process.argv);
